### DISCUSSION TITLE
why we need to add .env.* to the repository

### DISCUSSION THREAD

**Discussion by ikhattab** [vercel/next.js#22143] [Help] [answered]
I don't know why docs recommend adding `.env.*` to the repository; I believe it's not a good practice, as mentioned in https://12factor.net/config

<img width="692" alt="Screen Shot 2021-02-13 at 5 56 26 AM" src="https://user-images.githubusercontent.com/401304/107841513-177cbb00-6dc4-11eb-9b46-6d70bae82a46.png">

**Comment by michrome** [comment]
From [the Next.js docs](https://nextjs.org/docs/basic-features/environment-variables#default-environment-variables) :

> In general only one `.env.local` file is needed. However, sometimes you might want to add some defaults for the `development` (`next dev`) or `production` (`next start`) environment.

My reading of this is that is you might to commit environment specific defaults but **never secrets** (e.g. credentials), which is in agreement with https://12factor.net/config :

> The twelve-factor app stores config in environment variables (often shortened to env vars or env).

and

> A litmus test for whether an app has all config correctly factored out of the code is whether the codebase could be made open source at any moment, without compromising any credentials.

**Reply by ikhattab** [reply]
> My reading of this is that is you might to commit environment specific defaults but never secrets (e.g. credentials), which is in agreement with 12factor.net/config :

I don't think so, same paragraph stating clearly config values shouldn't be committed, environment variables aren't the same as config files. actually, the rest of the paragraph recommends using env variables instead of storing in config files that may easily be committed. security isn't the only concern of doing so

> The twelve-factor app stores config **in environment variables** (often shortened to env vars or env). Env vars are easy to change between deploys without changing any code; unlike config files, there is little chance of them being checked into the code repo accidentally; and unlike custom config files, or other config mechanisms such as Java System Properties, they are a language- and OS-agnostic standard.

you can also check https://github.com/motdotla/dotenv#faq
![image](https://user-images.githubusercontent.com/401304/107857390-d07bdd80-6e36-11eb-9b7e-b1d539e6df63.png)

**Reply by michrome** [reply]
[`<bikeshedding>`](https://en.wiktionary.org/wiki/bikeshedding)
It’s completely reasonable for a team to agree not to commit `.env.development`etc. A compelling reason might be to reduce the risk of someone accidentally committing a secret.

If the team makes that decision, they’ll likely have to agree how and where to document config and default values outside of the code.
`</bikeshedding>`

**Comment by timneutkens** [comment, marked as answer]
The dotenv loading order is based on the ruby dotenv package, which has the most widespread usage: https://github.com/bkeepers/dotenv#what-other-env-files-can-i-use

The `.local` variants should be gitignored, the others are for setting defaults.

**Reply by ikhattab** [reply]
I got the point; things become much clearer when I see the table. 

If **defaults** change from one environment to another that we need to commit the`.env.*` file for this environment; the points discussed in the https://12factor.net/config would still be of concern.

**Reply by ikhattab** [reply]
I believe the current setup is flexible to allows both approaches. Maybe this is only my own concern, and I can still not commit `.env.*` files if I want.

have a great day
